### PR TITLE
Python 3 and PyQt5 Support

### PR DESCRIPTION
**Pull request by m-kuhn**
At least some code from my side from Gran Canaria :)

This only includes build system and C++ changes.
Python files will still need to be updated. But there's a compatibility layer for PyQt to make code work with old and new version.

This feature was brought to you by OPENGIS.ch R&D

### CODE REVIEW

**Comment by mhugo** [comment]
Nice ! So there is a new compilation option (ENABLE_QT5) that switches the SIP bindings from Python2 to Python3, correct ?

**Comment by m-kuhn** [comment]
Correct

**Comment by 3nids** [comment]
Good job :)
Would be nice to merge it quickly, so we can properly test that the migration to python 3 doesn't affect the next release.
I'm not huge fan of merging at feature freeze, cool to see this now 
:+1:

**Comment by m-kuhn** [comment]
Instructions for building on ubuntu 15.10:

```
apt-get install \
                    bison \                                                     
                    cmake\                                                      
                    doxygen\                                                    
                    flex\                                                       
                    git\                                                        
                    graphviz\                                                   
                    grass-dev\                                                  
                    libexpat1-dev\                                              
                    libfcgi-dev\                                                
                    libgdal-dev\                                                
                    libgeos-dev\                                                
                    libgsl0-dev\                                                
                    libpq-dev\                                                  
                    libproj-dev\                                                
                    libspatialindex-dev\                                        
                    libspatialite-dev\                                          
                    libsqlite3-dev\                                             
                    lighttpd\                                                   
                    locales\                                                    
                    pkg-config\                                                 
                    poppler-utils\                                              
                    spawn-fcgi\                                                 
                    txt2tags\                                                   
                    xauth\                                                      
                    xfonts-100dpi\                                              
                    xfonts-75dpi\                                               
                    xfonts-base\                                                
                    xfonts-scalable\                                            
                    xvfb\                                                       
                    cmake-curses-gui\                                           
                    qttools5-dev\                                               
                    qttools5-dev-tools\                                         
                    libqt5svg5-dev\                                             
                    qtpositioning5-dev\                                         
                    qtscript5-dev\                                              
                    libqt5xmlpatterns5-dev\                                     
                    libqt5webkit5-dev\                                          
                    python3-pyqt5.qsci\                                         
                    libqca-qt5-2-dev\                                           
                    python3-sip-dev\                                            
                    libqwt-qt5-dev\                                             
                    pyqt5-dev-tools\                                            
                    libqt5scintilla2-dev
```

    cmake -DENABLE_QT5=ON -DWITH_QWTPOLAR=OFF ../QGIS
    make
    enjoy

**Comment by m-kuhn** [comment]
From my point of view, major issues should be sorted out by now. On the Qt4/Python2 case there should be no considerable effects.

The big advantage of merging this soon is, that it will allow us to collect experience with PyQt5 and Python3.

File: python/core/__init__.py [hunk 1]
```diff
@@ -139,7 +139,12 @@ def __hash__(self):
     NULL = QPyNullVariant(int)
 
 except ImportError:
-    pass
+    try:
+        # TODO: Fixme, this creates an invalid variant, not a NULL one
```
Comment (nyalldawson) [inline python/core/__init__.py]: Should this be resolved prior to merge?

Comment (m-kuhn) [inline python/core/__init__.py]: Not really a blocker. The major part of the python code isn't updated yet anyway and this one does not affect PyQt4 installations.

Comment (nyalldawson) [inline python/core/__init__.py]: Ok. Wasn't sure if it was a generic "sometime later" to-do or an overlooked pre-merge one.

**Comment by nirvn** [comment]
Nice job! 

However, on my Ubuntu 15.10 system, the build unfortunately segfaults upon startup. See gdb output:
[gdb-where.txt](https://github.com/qgis/QGIS/files/31526/gdb-where.txt)

**Comment by nirvn** [comment]
(the crash has to do with the python3-pyqt4 package. I could run the build until I installed that package.

**Comment by m-kuhn** [comment]
Did you start with a clean build directory?

**Comment by nirvn** [comment]
@m-kuhn roger that. Removing python3-pyqt4 does prevent the segault, and get me to a functional build, wouhou!

The python error box appears to be linked to a core component, not a plugin:


Traceback (most recent call last):
  File "", line 1, in 
  File "/home/webmaster/dev/cpp/QGIS/bm-qt5/output/python/qgis/utils.py", line 31, in 
    from PyQt4.QtCore import QCoreApplication, QLocale
ImportError: No module named 'PyQt4'

**Comment by m-kuhn** [comment]
@nirvn This is expected. Touching this code is slightly more dangerous than the changes proposed in this PR. But it will be the next step.

**Comment by nirvn** [comment]
@m-kuhn I've been relying on a qt5 build based on this PR for the last couple of days, works nicely.

Is the plan to merge this PR onto master soon? That'd be nice. 

BTW, I still get the keyboard shortcut issue with composer windows under Ubuntu 15.10 (qt5.4).

**Comment by m-kuhn** [comment]
Yeah, wanted to put it into quarantine for 7 days here and then merge if nobody objects.

**Comment by m-kuhn** [comment]
Merged 842c17c10b2112b0d92ab287d76e374900736065

**Pull request closed without merging**